This phylogeny shows evolutionary relationships of SARS-CoV-2 viral genomes from COVID-19 cases in Arizona, USA. Although the genetic relationships among sampled viruses are robust, there is considerable uncertainty surrounding estimates of transmission dates and in reconstruction of geographic spread. Please be aware that specific inferred transmission patterns are, at this time, only hypothesized.

Site numbering and genome structure uses [Wuhan-Hu-1/2019](https://www.ncbi.nlm.nih.gov/nuccore/MN908947) as reference. The phylogeny is rooted relative to early samples from Arizona. Temporal resolution assumes a nucleotide substitution rate of 8 Ã— 10^-4 subs per site per year.

Phylogenetic context of nCoV in SARS-related betacoronaviruses can be seen [here](https://nextstrain.org/groups/blab/sars-like-cov).

We gratefully acknowledge the diagnostic, sequencing, and submitting laboratories of the genetic sequences and metadata made available through [GISAID](https://gisaid.org) on which this research is based.

More information about the variants of SARS-CoV-2 circulating in Arizona can be found in our [web dashboard](https://pathogen.tgen.org/covidseq-tracker/).

Phylogenetic context of nCoV in SARS-related betacoronaviruses can be seen [here](https://nextstrain.org/groups/blab/sars-like-cov)
